Add bits to canvas_dump()

brianbienvenu opened the pull request:
Adds smaller text with the number of bits next to each BitField in a packet:
![image](https://user-images.githubusercontent.com/8326175/44153141-bd4c74b6-a0ea-11e8-92a6-2cfd12202aa8.png)

I think `\scriptsize` is about right visually

Will still go wider than the space set aside for it, if the field name is very long (try it with a field named `arbitrarily-long-field-name`). I don't think that's a big issue.

I also changed `fname` to `field`, as that's what it actually is. To print the field name, we were calling `fname.name` which doesn't make sense.

gpotter2 commented:
That’s really good job ! 👏

p-l- commented:
Maybe the byte size of other fields could be of interest?

gpotter2 commented:
Good idea, maybe in a different font and without the `b` so that it seems obvious

brianbienvenu commented:
I've added the bit/byte readout as a new column, including a left angle bracket to indicate little endian. Big (or "network") endian is the presumed default, to reduce visual noise. I tried with a `<` symbol, but it was unnecessarily wide.

The new column is bold, at 60% size and right aligned. Little endianness is inferred from having "LE" in the first three characters of the class name - I can't see a better way of doing that.

The value column has moved 0.5 units right to make space for the new column. I don't think this will interfere with the hex value dump table.

![image](https://user-images.githubusercontent.com/8326175/44319214-d9fe7780-a47d-11e8-9fe9-d55127d9e177.png)

Scapy's PyX is using the TeX environment, which makes everything a little harder than it should be (eg, `\textless` and '\texttt' doesn't exist). Changing PyX to LaTeX mode breaks the hexadecimal packet dump, which I don't really want to delve into at the moment. I ended up using a math environment to produce the left angle bracket.

I've also fixed a bug in the packet class name display - adding a `tex_escape` around it, allowing packet names to have underscores.

brianbienvenu commented:
@p-l- @gpotter2 do these changes seem OK to you?

gpotter2 commented:
I really like what you did, and love the new format.

The only thing I don’t like is the `<` symbol formatting, too “squashed”.

Also I don’t like having `LE` in the name column, because it seems that the field is called this way.

Sorry to be picky, and thanks again for your work

brianbienvenu commented:
Sorry for the unclear demo - the "LE" used to determine the presence of `<` in the bit/byte width field is in the class name, not the field name. I just had "LE" in the field name informationally. This is the code:

                if (hasattr(field, 'field') and
                        'LE' in field.field.__class__.__name__[:3] or
                        'LE' in field.__class__.__name__[:3]):

I agree on the shape of the `<`. I'll have another poke around on that.

gpotter2 commented:
Hi !
Any news on this ? I like this PR a lot and would like to get it merged. I think the `<` is quite OK already.
If you can improve it, that’s better but if you don’t have time we can merge it as-it

brianbienvenu commented:
Please merge as is, I haven't had time to get back to it. I will make another PR if I find time.

guedou commented:
@gpotter2 looks good to me but I cannot test it. You can merge it if it works fine =)
